src/shaders/shaders.js

```diff
@@ -102,12 +102,13 @@ function compile(fragmentSource, vertexSource) {
 varying ${precision} ${type} ${name};
 #else
 uniform ${precision} ${type} u_${name};
+${precision} ${type} ${name};
```

Comment:
I'm a bit concerned about changing variable lifetime as such and always using globals by default, mainly for register occupancy. Having tighter scope and keeping them localized can help the driver in determining when the variable does not warrant register slot usage. I haven't checked the effect of this change but this type of thing can be determined in renderdoc.

Would a workaround on your original issue of "would not be exposed to GLSL functions" to manually be passing these variables to the functions work for you?

Reply:
> Having tighter scope and keeping them localized can help the driver [...]

I have enough trust in compilers that this change is not an issue.

I doubt this leads to excessive register spilling, and I also doubt that there's a penalty for accessing global registers, even if the compiler *did* take the code at face value.
Even then, I doubt that it has a practical performance impact.

Ideally, we'd also find a better solution for shader optimization to avoid these concerns altogether (also see my comment in https://github.com/mapbox/mapbox-gl-js/issues/9390#issuecomment-728817028).

> [...] this type of thing can be determined in renderdoc.

I'm on macOS, which is currently unsupported by renderdoc (see https://github.com/baldurk/renderdoc/issues/1272).
I also have issues with the Imagination PVR tools at the moment, to check their disassembly (bug reported, but haven't gotten a response yet).

I'm also not sure how to debug mapbox-gl-js in those tools, and mapbox-gl-native is not upgraded for the latest mapbox-gl-js shaders (see https://github.com/mapbox/mapbox-gl-native/pull/16551) and doesn't work with latest macOS toolchains (see https://github.com/mapbox/mapbox-gl-native/pull/16550). I'm also about to give up trying to upstream to mapbox-gl-native due to slow review.

That said, I'm not sure how renderdoc could be used to analyze the shader-compiler behavior, and what could be derived about practical effects.

> Would a workaround on your original issue of "would not be exposed to GLSL functions" to manually be passing these variables to the functions work for you?

Yes, but this issue is more about making the code more consistent.
Passing them manually (or using the `#pragma` in each function) would also have an impact on my tooling which generates my shader code. I'd rather avoid such hacks.

Manually adding copies of the variable declaration also increases the risk of using the wrong type / precision (where it's only changed in one spot, but forgotten elsewhere). Adding it as function argument might *also* affect performance in other ways.

Another alternative strategy was to use the preprocessor (`#define ${name} u_${name}`), but I was concerned with running into issues with the mapbox minifier or some other edge-case with different preprocessor behaviour.


Just to be clear: I'll keep this code in our fork, and I'm fine if this change is rejected upstream (here) - it won't affect *me*.
I just think that it helps readability of the code and that it avoids errors by making it more consistent *for everyone*.

Reply:
>I doubt this leads to excessive register spilling, and I also doubt that there's a penalty for accessing global registers, even if the compiler did take the code at face value.
Even then, I doubt that it has a practical performance impact.

In doubt, I'd expect the worst. There is a risk to changing all these shader variables as such. A few example of such issues from this project: https://github.com/mapbox/mapbox-gl-js/issues/5490, https://github.com/mapbox/mapbox-gl-js/issues/9327, https://github.com/mapbox/mapbox-gl-js/issues/9370, https://github.com/mapbox/mapbox-gl-js/issues/4679... and these were visual regressions, performance may be subtle/platform specific and go unnoticed. In those issues, some of the related changes that introduced those regressions were rather minimal.

Thanks a lot for taking the time on putting together this PR and description, but the benefits of this change are really for the fork you are maintaining rather than for a wider use of the library itself; I'd rather limit such risk by not introducing this change. For the purpose of your issue, I would go with the proposed workaround.

Reply:
Concerning renderdoc, we usually make captures on chrome windows using --gpu-startup-dialog, --disable-gpu-sandbox and --disable-gpu-watchdog, http://shader-playground.timjones.io/ may also be useful.